contracts/src/libraries/Transcript.sol

```diff
@@ -60,15 +60,19 @@ library Transcript {
     }
 
     function getAndAppendChallenge(TranscriptData memory self) internal pure returns (uint256) {
-        bytes32 h1 =
-            keccak256(abi.encodePacked(self.state[0], self.state[1], self.transcript, uint8(0)));
-        bytes32 h2 =
-            keccak256(abi.encodePacked(self.state[0], self.state[1], self.transcript, uint8(1)));
+        bytes32 hash = keccak256(abi.encodePacked(self.state, self.transcript));
 
-        self.state[0] = h1;
-        self.state[1] = h2;
+        self.state = hash;
 
-        return BN254.fromLeBytesModOrder(BytesLib.slice(abi.encodePacked(h1, h2), 0, 48));
+        // TODO Philippe create function in BN254.sol
+        // TODO Optimize
+        bytes memory beBytes = abi.encodePacked(hash);
+        uint256 ret = 0;
+        for (uint256 i = 0; i < beBytes.length; i++) {
+            ret = mulmod(ret, 256, BN254.R_MOD);
+            ret = addmod(ret, uint256(uint8(beBytes[i])), BN254.R_MOD);
+        }
+        return ret;
```

Comment:
personally I find this to be gas inefficient. since we know for sure that `beBytes` will be 32 bytes, (first of all, i don't know why we convert it to a generic `bytes` from `bytes32`)

is the following wrong?:
```solidity
uint256 ret = uint256(hash) % BN254.R_MOD;
```

Reply:
Thanks, you are right. I fixed it in @imabdulbasit 's branch: 9cb190cc
With this change verifying the proof now costs 600k instead of 757k!!! Very exciting when this change is merged with the other optimizations :)

Reply:
great to hear!